File: CPE/Star/3/UVA-11407_Squares.cpp
/*
    Dynamic Programming
    dp[i] 表示和為i的完全平方數的最小個數
*/
#include <bits/stdc++.h>
using namespace std;
const int N = 1e4 + 5;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    vector<int> dp(N);
    for (int i = 1; i < N; ++i) {
        dp[i] = i;
        for (int j = 1; j * j <= i; ++j) {
            dp[i] = min(dp[i], dp[i - j * j] + 1);
        }
    }
    int t, n;
    cin >> t;
    while (t--) {
        cin >> n;
        cout << dp[n] << endl;
    }
    return 0;
}

File: CPE/Star/3/UVA-11407_Squares.py
"""
    Dynamic Programming + 預處理
    dp[i] 表示和為i的完全平方數的最小個數
"""
N = 10000
t = int(input())
dp = [0] * (N + 1)
for i in range(1, N + 1):
    dp[i] = i # i = (1^2) * i
    for j in range(1, int(i ** 0.5) + 1): # j*j <= i
        dp[i] = min(dp[i], dp[i - j * j] + 1)

for _ in range(t):
    n = int(input())
    print(dp[n])